### PR TITLE
[#178629709] Snapshots research

### PR DESCRIPTION
**Pull request by aquiladev**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -315,3 +315,141 @@ Changes:
...
+
+### Example of usage
+
+```
+const Sandbox = require('./sandbox');
+
+describe('Test', async () => {
+  before(async () => {
+    sandbox = await Sandbox.create();
+    await sandbox.start();
```
Comment (bogdan) [inline README.md:426]: It would be cool to have `sandbox = await Sandbox.start();` - make it kinda easier to use.

File: uns-config.json [hunk 2]
```diff
@@ -139,6 +139,47 @@
                     "deprecated": true
                 }
             }
+        },
+        "1337": {
+            "contracts": {
+                "CNSRegistry": {
+                    "address": "0xA310F8e781F8aE80690649be920991AEc9a6595D",
+                    "deploymentBlock": "0x01"
```
Comment (bogdan) [inline uns-config.json:147]: Roman had reported that his JSON parser has problems if some field is absent for some contract. We need to add `legacyAddresses: []` for every contract and make sure all fields are specified everywhere.

File: uns-config.json [hunk 3]
```diff
@@ -139,6 +139,47 @@
...
+                "MintingController": {
+                    "address": "0x4513073d9B3af7e62be59B57205a2368F76dE9C7",
+                    "deploymentBlock": "0x03"
+                },
+                "URIPrefixController": {
+                    "address": "0xeE1f42382091e260f5B9c152E17639d729Ab829F",
+                    "deploymentBlock": "0x04"
+                },
+                "Resolver": {
+                    "address": "0x76CBBE3B425A7D258B30457785d54de8c8322bfa",
```
Comment (bogdan) [inline uns-config.json:162]: I remember there were some tests in website that deployed a legacy resolver and used `Resolver.legacyAddresses` field. That might be an issue, so it is possible we have to deploy a legacy resolver too.

File: sandbox/index.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  totalAccounts: 10,
+  hardfork: 'petersburg',
+  allowUnlimitedContractSize: false,
+  locked: false,
+  hdPath: 'm/44\'/60\'/0\'/0/',
+  keepAliveTimeout: 5000,
+  mnemonic: 'mimic dune forward party defy island absorb insane deputy obvious brother immense',
+  chainId: 1337,
+  dbPath: './.sandbox',
+  snapshotPath: './sandbox/db.tgz',
```
Comment (bogdan) [inline sandbox/index.js:20]:
```suggestion
  snapshotPath: require('path').join(__dirname, './sandbox/db.tgz'),
```

File: package.json [hunk 5]
```diff
@@ -9,9 +9,12 @@
     "compile": "hardhat compile",
     "compile:size": "env ENABLE_CONTRACT_SIZER=true yarn compile",
     "coverage": "hardhat coverage",
-    "test": "hardhat test",
-    "lint": "npm run lint:js && npm run lint:sol",
-    "lint:fix": "npm run lint:js:fix && npm run lint:sol:fix",
+    "test": "yarn test:sol && yarn test:sandbox",
+    "test:sol": "hardhat test",
+    "test:sandbox": "env HARDHAT_NETWORK=sandbox mocha ./sandbox/**/*.test.js",
+    "rebuild:sandbox": "env HARDHAT_NETWORK=sandbox yarn node ./sandbox",
```
[the remark this answers is not in the document]
Comment (aquiladev) [reply package.json]: all scripts build with name {action}:{what}

File: src/snapshot.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  return await network.provider.request({
+    method: 'evm_revert',
+    params: [snapshotId],
+  });
+}
+
+module.exports = {
+  snapshot,
+  revert,
+};
```
[the remark this answers is not in the document]
Comment (aquiladev) [reply src/snapshot.js:20]: only for tests, will remove it

File: sandbox/index.js [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+      if (e.message.includes('Server is not running')) {
+        GanacheService.error = undefined;
+        return;
+      }
+      throw e;
+    }
+  }
+
+  async reset () {
+    await this._revert(this.snapshotId);
```
[the remark this answers is not in the document]
Comment (aquiladev) [reply sandbox/index.js:70]: revert execution throws error

File: sandbox/index.js [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    this.snapshotId = await this._snapshot();
+  }
+
+  async rebuild () {
+    await this.start();
+    const deployer = await Deployer.create();
+    const deployConfig = await deployer.execute(['full']);
+    console.log('Config:', JSON.stringify(deployConfig));
```
[the remark this answers is not in the document]
Comment (aquiladev) [reply sandbox/index.js]: plan to replace it by debug

File: uns-config.json [hunk 9]
```diff
@@ -139,6 +139,47 @@
...
+                    "deploymentBlock": "0x08"
+                },
+                "ProxyAdmin": {
+                    "address": "0x7bB6Cd9be29fab783c0b494A06FED8b2E2596B7a"
```
[the remark this answers is not in the document]
Comment (aquiladev) [reply uns-config.json:166]: it is kinda difficult to find

**Pull request merged**